### PR TITLE
Add support for start and end date args, to allow more precision over lookback

### PR DESCRIPTION
**Pull request by joeyTedeschi**
Add support for `--start-date` and `--end-date`, to allow greater precision when pulling GH data.

Also renamed the `short-flag` for `--exclude-orgs` to `-x` to avoid overlap with `-e` for `--end-date`, but happy to use another letter 🤷

### CODE REVIEW

**Comment by bostonaholic** [comment]
Thanks for contributing to this! I've been wanting this feature for a while and haven't gotten to it: #2

File: README.md [hunk 1]
```diff
@@ -62,22 +62,31 @@ Example:
 
 **Required:**
 - `-u, --username <username>`: Your GitHub username to fetch activity for
+
+**One of the following date options is required:**
 - `-l, --lookback <number>`: Number of months to look back for activity (must be a positive number)
+- `-s, --start-date <date>`: Explicit start date in DD-MM-YYYY format, that must be used with --end-date
```
Comment (bostonaholic) [inline README.md:68]: Let's leave out the short flags for now. I want to avoid having to figure out which single letter is best for which flag. I will probably eventually remove all short flags sometime.

This is why I went with `lookback` when I originally had `months` but then that conflicted with `-m` for `--model`.

File: README.md [hunk 2]
```diff
@@ -62,22 +62,31 @@ Example:
...
+- `-e, --end-date <date>`: Explicit end date in DD-MM-YYYY format, that must be used with --start-date
 
 **Optional:**
 - `-p, --provider <provider>`: LLM provider to use (e.g., openai, anthropic), defaults to openai
 - `-m, --model <model>`: OpenAI model to use (e.g., gpt-4, gpt-3.5-turbo), defaults to gpt-4o
 - `-b, --brag`: Optional flag to generate a summary and brag document
 - `-i, --include-orgs <orgs...>`: Only include contributions from these organizations (mutually exclusive with --exclude-orgs)
-- `-e, --exclude-orgs <orgs...>`: Exclude contributions from these organizations (mutually exclusive with --include-orgs)
+- `-x, --exclude-orgs <orgs...>`: Exclude contributions from these organizations (mutually exclusive with --include-orgs)
```
Comment (bostonaholic) [inline README.md:76]: Avoid breaking changes for now until I'm able to remove all of the short flags.

```suggestion
- `-e, --exclude-orgs <orgs...>`: Exclude contributions from these organizations (mutually exclusive with --include-orgs)
```

File: README.md [hunk 3]
```diff
@@ -62,22 +62,31 @@ Example:
...
 
 ### Examples 🚀
 
-Basic usage:
+Basic usage with lookback:
 ```bash
 ./reflect --username bostonaholic --lookback 6 --brag
 ```
 
+Basic usage with explicit date range:
```
Comment (bostonaholic) [inline README.md:85]: Thanks for adding the example!

File: __tests__/unit/cli.test.ts [hunk 4]
```diff
@@ -99,4 +99,63 @@ describe('CLI Validation Functions', () => {
       );
     });
   });
-});
\ No newline at end of file
+
+  describe('isValidDateFormat', () => {
```
Comment (bostonaholic) [inline __tests__/unit/cli.test.ts:103]: Great job writing the unit tests!

**suggestion (non-blocking):** Try writing these as property-based tests using [fast-check](https://www.npmjs.com/package/fast-check).

If you're not familiar with property-based testing, don't worry about it. I can include them after this is merged.

tl;dr Property-based tests are a way to test your code by describing general properties or rules that should always hold true, no matter what input is given. Instead of testing specific examples (like unit tests), they generate many random inputs automatically and check that the function behaves correctly for all of them. This can help find edge cases or unexpected bugs that might not be obvious when writing individual test cases.

For example, with isValidDateFormat, instead of testing a few hardcoded dates like "2025-04-05" or "invalid-date", we could use property-based testing to automatically generate hundreds (or thousands!) of random strings and verify that the function correctly identifies valid and invalid formats. This approach complements your existing unit tests by covering a much broader range of inputs.

Let me know if you'd like help getting started with this!

File: lib/cli.ts [hunk 5]
```diff
@@ -94,6 +141,44 @@ function validateProvider(provider: LlmProvider): void {
...
+  }
+
+  const formattedStartDate = new Date(startDate);
+  const formattedEndDate = new Date(endDate);
+  
+  validateDateRange(formattedStartDate, formattedEndDate);
+}
+
+
+function validateDateRange(startDate: Date, endDate: Date): void {
```
Comment (bostonaholic) [inline lib/cli.ts:162]: Nice job on these validations!

File: lib/date-utils.ts [hunk 6]
```diff
@@ -1,7 +1,34 @@
-export function calculateDateRange(months: number): { startDate: Date; endDate: Date } {
+import chalk from "chalk";
+
+export function calculateDateRange(lookbackMonths?: number, startDate?: string, endDate?: string): { startDate: Date; endDate: Date } {
+  if (startDate && endDate) {
+    if (lookbackMonths) {
+      console.warn(chalk.yellow('! Lookback months will be ignored when startDate or endDate are provided'));
```
Comment (bostonaholic) [inline lib/date-utils.ts:6]:
```suggestion
      console.warn(chalk.yellow('! Lookback months will be ignored when startDate and endDate are provided'));
```

File: lib/date-utils.ts [hunk 7]
```diff
@@ -1,7 +1,34 @@
...
+    }
+
+    // Looking to avoid issue where it gets converted to local time, and changes date
+    const parsedStartDate = new Date(`${startDate}T00:00:00`);
+    const parsedEndDate = new Date(`${endDate}T23:59:59`);
+
+    console.log(chalk.green('Pulling github data from'), chalk.cyan(parsedStartDate), chalk.green('to'), chalk.cyan(parsedEndDate));
```
Comment (bostonaholic) [inline lib/date-utils.ts:13]: **suggestion (non-blocking):** Put the date range in the `spinner` in `fetchMergedPRs` and `fetchClosedIssues` so it says something like:

`Fetching merged pull requests between {startDate} and {endDate}...`

File: lib/date-utils.ts [hunk 8]
```diff
@@ -1,7 +1,34 @@
...
+  // Likely won't be hit as this should be validated by cli, but including for completeness
+  console.error(chalk.red('✖ Error: Either --lookback, or --start-date and --end-date, must be provided'));
+  process.exit(1);
+}
+
+function calculateLookbackDateRange(lookbackMonths: number): { startDate: Date; endDate: Date } {
   const endDate = new Date();
   const startDate = new Date();
-  startDate.setMonth(startDate.getMonth() - months);
+  startDate.setMonth(startDate.getMonth() - lookbackMonths);
```
Comment (bostonaholic) [inline lib/date-utils.ts:31]: Oh, cool! I hadn't thought of doing it like this.

`--end-date 2024-12-31 --lookback 6`

I like it!
